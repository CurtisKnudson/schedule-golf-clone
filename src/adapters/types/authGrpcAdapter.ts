// Protos
import {
  CreateNewUserResponse,
  UserLoginResponse,
} from 'gen/proto/ts/schedule_golf/authentication/v1alpha1/authentication';

// Types
import { NewUser } from 'types/user';
import { LoginCredentials } from 'pages/authentication/types/loginCredentials';

export interface AuthGrpcAdapter {
  createNewUser(newUser: NewUser): Promise<CreateNewUserResponse>;
  userLogin(userLoginCredentials: LoginCredentials): Promise<UserLoginResponse>;
  userTokenRefresh(): Promise<{
    isAuthenticated: boolean;
  }>;
}
