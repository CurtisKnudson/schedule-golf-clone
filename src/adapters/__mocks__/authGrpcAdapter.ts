// Adapter
import { AuthGrpcAdapter as AuthGrpcAdapterInterface } from 'adapters/types/authGrpcAdapter';
// Protos
import {
  CreateNewUserRequest,
  UserLoginRequest,
  UserLoginResponse,
} from 'gen/proto/ts/schedule_golf/authentication/v1alpha1/authentication';

// Types
import { NewUser } from 'types/user';
import { LoginCredentials } from 'pages/authentication/types/loginCredentials';

export class AuthGrpcAdapterMock implements AuthGrpcAdapterInterface {
  async createNewUser(newUser: NewUser) {
    const req: CreateNewUserRequest = {
      ...newUser,
    };

    return Promise.resolve(req);
  }

  async userLogin(userLoginCredentials: LoginCredentials) {
    if (!userLoginCredentials) {
      throw new Error('You must provide user login credentials');
    }
    const req: UserLoginRequest = {
      ...userLoginCredentials,
    };

    const res: UserLoginResponse = {
      userId: '123456789',
      companyName: 'Riverbend Golf Community',
      firstName: 'Curtis',
      lastName: 'Knudson',
      email: req.email,
    };
    return Promise.resolve(res);
  }

  async userTokenRefresh() {
    return Promise.resolve({
      isAuthenticated: true,
    });
  }
}
