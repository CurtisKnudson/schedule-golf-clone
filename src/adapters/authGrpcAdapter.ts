// Node Modules
import { GrpcWebFetchTransport } from '@protobuf-ts/grpcweb-transport';
// Types
import { LoginCredentials } from 'pages/authentication/types/loginCredentials';
import { NewUser } from 'types/user';
import {
  AuthGrpcAdapter as AuthGrpcAdapterInterface,
  RefreshResponse,
} from 'adapters/types/authGrpcAdapter';
import { GrpcCode } from 'types/code';
// Constants
import { apiUrl } from 'constants/apiUrl';
// Protos
import {
  CreateNewUserRequest,
  ForeUpAuthenticationRequest,
  UserLoginRequest,
  UserTokenRefreshRequest,
} from 'gen/proto/ts/schedule_golf/authentication/v1alpha1/authentication';
import { AuthenticatorServiceClient } from 'gen/proto/ts/schedule_golf/authentication/v1alpha1/authentication.client';
// Utils
import { getCookie, setCookie } from 'utils/cookieHelper';

export class AuthGrpcAdapter implements AuthGrpcAdapterInterface {
  transport = new GrpcWebFetchTransport({
    baseUrl: apiUrl,
  });
  authenticator = new AuthenticatorServiceClient(this.transport);

  async createNewUser(newUser: NewUser) {
    const req: CreateNewUserRequest = {
      ...newUser,
    };

    const res = await this.authenticator.createNewUser(req);

    if (
      typeof res.headers.token !== 'string' ||
      typeof res.headers.expiration !== 'string'
    ) {
      throw new Error('Request Headers returned an invalid token or expiration');
    }

    // Need to set a cookie that expires based on the return value in the metadata
    setCookie('token', res.headers.token, res.headers.expiration);

    return {
      res: res.response,
      expiration: res.headers.expiration,
    };
  }

  async userLogin(userLoginCredentials: LoginCredentials) {
    const req: UserLoginRequest = {
      ...userLoginCredentials,
    };

    const res = await this.authenticator.userLogin(req);

    if (
      typeof res.headers.token !== 'string' ||
      typeof res.headers.expiration !== 'string'
    ) {
      throw new Error('Request Headers returned an invalid token or expiration');
    }

    setCookie('token', res.headers.token, res.headers.expiration);

    return {
      res: res.response,
      expiration: res.headers.expiration,
    };
  }

  async userTokenRefresh() {
    const jwt = getCookie('token');

    if (!jwt) {
      throw new Error('userTokenRefresh did not receive a webtoken');
    }

    const req: UserTokenRefreshRequest = {
      jwt,
    };

    const res = await this.authenticator.userTokenRefresh(req);

    const { status } = res.response;

    const refreshResponse: RefreshResponse = {
      isAuthenticated: false,
      expiration: null,
    };

    // If JWT is succesfully refreshed, update the cookie
    if (status?.code === GrpcCode.OK) {
      setCookie('token', res.response.jwt, res.response.expiration);
      refreshResponse.isAuthenticated = true;
      refreshResponse.expiration = res.response.expiration;
      return refreshResponse;
    }
    // If the JWT is cancelled by the API, leave everything as it is. This means that the JWT is still valid
    if (status?.code === GrpcCode.CANCELLED) {
      refreshResponse.isAuthenticated = true;
      return refreshResponse;
    }

    return refreshResponse;
  }

  async foreUpAuthentication({
    email,
    password,
    scheduleGolfJwt,
  }: ForeUpAuthenticationRequest) {
    const req: ForeUpAuthenticationRequest = {
      email,
      password,
      scheduleGolfJwt,
    };

    const res = await this.authenticator.foreUpAuthentication(req);

    console.log(res);
  }
}
