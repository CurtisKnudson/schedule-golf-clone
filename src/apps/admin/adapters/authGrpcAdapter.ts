// Node Modules
import { GrpcWebFetchTransport } from '@protobuf-ts/grpcweb-transport';
// Types
import { LoginCredentials } from 'apps/admin/types/loginCredentials';
import { NewUser } from 'apps/admin/types/user';
import {
  AuthGrpcAdapter as AuthGrpcAdapterInterface,
  RefreshResponse,
} from 'apps/admin/adapters/types/authGrpcAdapter';
import { GrpcCode } from 'types/googleCodes';
// Constants
import { apiUrl } from 'apps/admin/constants/apiUrl';
// Protos
import {
  CreateNewUserRequest,
  ForeUpAuthenticationRequest,
  UserLoginRequest,
  UserTokenRefreshRequest,
} from 'gen/proto/ts/schedule_golf/authentication/v1alpha1/authentication';
import { AuthenticatorServiceClient } from 'gen/proto/ts/schedule_golf/authentication/v1alpha1/authentication.client';
// Utils
import { getCookie, setCookie } from 'utils/cookieHelper';

export class AuthGrpcAdapter implements AuthGrpcAdapterInterface {
  transport = new GrpcWebFetchTransport({
    baseUrl: apiUrl,
  });
  authenticator = new AuthenticatorServiceClient(this.transport);

  async createNewUser(newUser: NewUser) {
    const req: CreateNewUserRequest = {
      ...newUser,
    };

    const res = await this.authenticator.createNewUser(req);

    if (
      typeof res.headers.token !== 'string' ||
      typeof res.headers.expiration !== 'string'
    ) {
      throw new Error('Request Headers returned an invalid token or expiration');
    }

    // Need to set a cookie that expires based on the return value in the metadata
    setCookie('token', res.headers.token, res.headers.expiration);

    return {
      res: res.response,
      expiration: res.headers.expiration,
    };
  }

  async userLogin(userLoginCredentials: LoginCredentials) {
    const req: UserLoginRequest = {
      ...userLoginCredentials,
    };

    const res = await this.authenticator.userLogin(req);

    if (
      typeof res.headers.token !== 'string' ||
      typeof res.headers.expiration !== 'string'
    ) {
      throw new Error('Request Headers returned an invalid token or expiration');
    }

    setCookie('token', res.headers.token, res.headers.expiration);

    return {
      res: res.response,
      expiration: res.headers.expiration,
    };
  }

  async userTokenRefresh() {
    const jwt = getCookie('token');
    const refreshResponse: RefreshResponse = {
      isAuthenticated: false,
      expiration: null,
    };

    if (!jwt) {
      return refreshResponse;
    }

    const req: UserTokenRefreshRequest = {
      jwt,
    };

    const res = await this.authenticator.userTokenRefresh(req);

    const { status } = res.response;

    // If JWT is succesfully refreshed, update the cookie
    if (status?.code === GrpcCode.OK) {
      setCookie('token', res.response.jwt, res.response.expiration);
      refreshResponse.isAuthenticated = true;
      refreshResponse.expiration = res.response.expiration;
      return refreshResponse;
    }
    // If the JWT is cancelled by the API, leave everything as it is. This means that the JWT is still valid
    if (status?.code === GrpcCode.CANCELLED) {
      refreshResponse.isAuthenticated = true;
      return refreshResponse;
    }

    return refreshResponse;
  }

  async foreUpAuthentication({
    email,
    password,
    scheduleGolfJwt,
  }: ForeUpAuthenticationRequest) {
    const req: ForeUpAuthenticationRequest = {
      email,
      password,
      scheduleGolfJwt,
    };

    const res = await this.authenticator
      .foreUpAuthentication(req)
      .then((res) => {
        return {
          isSucces: true,
          message: res.response.status?.message
            ? res.response.status?.message
            : 'success',
        };
      })
      .catch(() => {
        return {
          isSuccess: false,
          message: '',
        };
      });

    return res;
  }
}
