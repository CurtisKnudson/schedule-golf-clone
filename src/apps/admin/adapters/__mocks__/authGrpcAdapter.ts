// Adapter
import { AuthGrpcAdapter as AuthGrpcAdapterInterface } from 'apps/admin/adapters/types/authGrpcAdapter';
// Protos
import {
  CreateNewUserRequest,
  ForeUpAuthenticationRequest,
  UserLoginRequest,
  UserLoginResponse,
} from 'gen/proto/ts/schedule_golf/authentication/v1alpha1/authentication';

// Types
import { NewUser } from 'types/user';
import { LoginCredentials } from 'apps/admin/types/loginCredentials';

export class AuthGrpcAdapterMock implements AuthGrpcAdapterInterface {
  async createNewUser(newUser: NewUser) {
    const req: CreateNewUserRequest = {
      ...newUser,
    };

    return Promise.resolve({
      res: req,
      expiration: new Date().toString(),
    });
  }

  async userLogin(userLoginCredentials: LoginCredentials) {
    if (!userLoginCredentials) {
      throw new Error('You must provide user login credentials');
    }
    const req: UserLoginRequest = {
      ...userLoginCredentials,
    };

    const res: UserLoginResponse = {
      userId: '123456789',
      companyName: 'Riverbend Golf Community',
      firstName: 'Curtis',
      lastName: 'Knudson',
      email: req.email,
    };
    return Promise.resolve({
      res,
      expiration: new Date().toString(),
    });
  }

  async userTokenRefresh() {
    return Promise.resolve({
      isAuthenticated: true,
      expiration: new Date().toString(),
    });
  }

  async foreUpAuthentication({
    email,
    password,
    scheduleGolfJwt,
  }: ForeUpAuthenticationRequest) {
    const req = {
      email,
      password,
      scheduleGolfJwt,
    };
    return {
      isSuccess: true,
      message: req.email,
    };
  }
}
