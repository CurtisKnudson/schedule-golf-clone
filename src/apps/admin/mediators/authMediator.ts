// Node Modules
import { v4 } from 'uuid';
// Types
import {
  AuthMediatorInterface,
  ForeUpAuthenticationRequest,
} from 'apps/admin/mediators/types/authMediatorInterface';
import { LoginCredentials } from 'apps/admin/types/loginCredentials';
import { NewUser } from 'types/user';
// Adapter
import { AuthGrpcAdapter } from 'apps/admin/adapters/types/authGrpcAdapter';

export class AuthMediator implements AuthMediatorInterface {
  private adapter: AuthGrpcAdapter;

  constructor(adapter: AuthGrpcAdapter) {
    this.adapter = adapter;
  }

  async createNewUser(newUser: NewUser) {
    if (!newUser) {
      throw new Error('You must provide newUser credentials');
    }
    newUser.userId = v4();
    const req = await this.adapter.createNewUser(newUser);

    // TODO: Need Error Handling
    // What if the user authentication fails?
    return req;
  }

  async userLogin(userLoginCredentials: LoginCredentials) {
    if (!userLoginCredentials) {
      throw new Error('You must provide login credentials to login');
    }
    if (!userLoginCredentials.email) {
      throw new Error('You must provide an email to login');
    }
    if (!userLoginCredentials.password) {
      throw new Error('You must provide a password to login');
    }

    const req = await this.adapter.userLogin(userLoginCredentials);

    return req;
  }
  async userTokenRefresh() {
    return await this.adapter.userTokenRefresh();
  }
  async foreUpAuthentication(authRequest: ForeUpAuthenticationRequest) {
    return await this.adapter.foreUpAuthentication(authRequest);
  }
  dispose() {
    return;
  }
}
