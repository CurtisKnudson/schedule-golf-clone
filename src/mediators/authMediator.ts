// Node Modules
import { AuthGrpcAdapter } from 'adapters/types/authGrpcAdapter';
// Types
import { AuthMediatorInterface } from 'mediators/types/authMediatorInterface';
import { NewUser } from 'types/user';
import { v4 } from 'uuid';

export class AuthMediator implements AuthMediatorInterface {
  private adapter: AuthGrpcAdapter;

  constructor(adapter: AuthGrpcAdapter) {
    this.adapter = adapter;
  }

  async createNewUser(newUser: NewUser) {
    if (!newUser) {
      throw new Error('You must provide newUser credentials');
    }
    newUser.userId = v4();
    const req = await this.adapter.createNewUser(newUser);

    // TODO: Need Error Handling
    // What if the user authentication fails?
    return req;
  }

  dispose() {
    return;
  }
}
