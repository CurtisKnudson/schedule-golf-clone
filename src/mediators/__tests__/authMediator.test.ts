// Node Modules
import { describe, it, expect } from 'vitest';
// Mocks
import { AuthGrpcAdapterMock } from 'adapters/__mocks__/authGrpcAdapter';
// Mediators
import { AuthMediator } from 'mediators/authMediator';

describe('Auth Mediator', () => {
  const adapter = new AuthGrpcAdapterMock();
  const authMediator = new AuthMediator(adapter);

  it('Create New User', async () => {
    const req = await authMediator.createNewUser({
      userId: '1234567890',
      companyName: 'Riverbend Golf Community',
      firstName: 'Curtis',
      lastName: 'Knudson',
      email: 'curtis@gmail.com',
      password: 'testpassword',
    });
    expect(req).toMatchObject({
      companyName: 'Riverbend Golf Community',
      firstName: 'Curtis',
      lastName: 'Knudson',
      email: 'curtis@gmail.com',
      password: 'testpassword',
    });
  });
});
