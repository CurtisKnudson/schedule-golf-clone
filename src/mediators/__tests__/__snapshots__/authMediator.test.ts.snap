// Vitest Snapshot v1

exports[`Auth Mediator > Create New User 1`] = `
{
  "companyName": "Riverbend Golf Community",
  "email": "curtis@gmail.com",
  "firstName": "Curtis",
  "lastName": "Knudson",
  "password": "testpassword",
  "userId": Any<String>,
}
`;
